RockSniffer/RPC/AlbumArtResolver.cs

```diff
@@ -0,0 +1,136 @@
+﻿using Newtonsoft.Json;
+using RockSnifferLib.Logging;
+using RockSnifferLib.Sniffing;
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+
+namespace RockSniffer.RPC
+{
+    public class AlbumArtResolver
+    {
+        [Serializable]
+        private class AppleResult
+        {
+            public string artistName;
+            public string collectionName;
+            public string trackName;
+            public string artworkUrl100;
+        }
+
+        [Serializable]
+        private class AppleRequestResponse
+        {
+            public int resultCount;
+            public AppleResult[] results;
+        }
+
+        private HttpClient httpClient = new HttpClient();
+        private Dictionary<string, (string URL, string DisplayText)> cache = new Dictionary<string, (string URL, string DisplayText)>();
+
+
+        public AlbumArtResolver() {
+            httpClient.DefaultRequestHeaders.Add("User-Agent", "RockSniffer");
+        }
+
+        /// <summary>
+        /// Attempt to obtain album cover
+        /// </summary>
+        /// <param name="songInfo">Details about the given song</param>
+        public (string URL, string DisplayText)? Get(SongDetails songInfo)
+        {
+            string key = $"{songInfo.artistName}|{songInfo.albumName}";
+
+            // Use double-checked locking pattern, since there are multiple threads
+            if (!cache.ContainsKey(key))
+                lock (this)
+                    if (!cache.ContainsKey(key))
+                    {
+                        //Logger.Log("AlbumArtResolver::Get : Cache miss on {0}", key);
+                        if (GetFromAppleMusic(songInfo) is (string, string) validResult)
+                        {
+                            cache[key] = validResult;
+                        }
+                    }
+
+            if (!cache.ContainsKey(key))
+            {
+                Logger.Log("AlbumArtResolver::Get : Could not find any album art for '{0}'", key);
```

Comment:
This seems to be printed around 5-10 times per second when playing a song where art is not found, for example `RapScallions - California Brain`. Store something in the cache to indicate that an attempt was made, but it was unsuccessful.

Reply:
Hmm, I was thinking it would be better not to save, in case it succeeds later, but I'm not so sure now, I'll think it through. However 5-10 times per seconds sounds like entirely too much: I assume that's how often the Discord status is being updated, which sounds, again, like entirely too much 🤔